### ISSUE TITLE
Simpler opening example for "git replace" in Chapter 7?

### ISSUE THREAD

**Issue by rpjday** [progit/progit2#909]
I've never used "git replace" before, and the example used in Chapter 7 seems unnecessarily complicated (possibly because I don't understand it yet :-), so I wanted to make sure I understand how "git replace" works in order to suggest a simpler, first example.

As I read it, git replace works like git notes -- there is no change to the underlying commit graph; rather, all changes come from outside as references into the commit graph. That part's pretty easy. So what about this as an example?

Assume you had some file buried in history that contained content that wasn't necessarily critical or security-sensitive, just some content that you preferred people not tripping over by accident. Of course, they could always see it if they availed themselves of "--no-replace-objects", but we understand that. All we want to do is make that content less readily available to the casual user.

Well, what about creating an alternate version of that file, then (using the same recipe with which you can [embed a public key in a repo](http://365git.tumblr.com/post/2813251228/adding-a-gpg-public-key-to-a-repository)), planting that new blob in the repo, and using it to "replace" the original version of the file? Is that not a viable thing to do?

As long as I'm properly understanding how git replace works, I think this would be an easier example to start with, and it can always be followed by what's in that section now.

Thoughts?

**Comment by ben** [comment]
This is a pretty esoteric command, and it's hard to come up with a simple use case that will make every reader think "yeah, I could see myself needing that."

The idea of using an orphan tag/blob to include "sensitive" information is kind of a neat trick, but what's the advantage over just including the `pubkey.txt` file in the commit-stream tree? It might be a tad messier, but it means you can avoid having to execute a plumbing command to get the key to verify the signatures.

In either case, the "sensitive" information isn't really protected. Public keys _should_ be public, and private keys have **no** place in a source code repository, so making some kind of in-the-middle state for either of them doesn't seem useful.

In my experience, replacements and grafts are useful while you're reworking the history of a project. (I once did this to migrate a project from a weird VCS to git: (1) snapshot the latest and freeze it into an initial commit, where the rest of the team can continue working from there; (2) import the 10-year history of the source code into Git over the course of several days; then (3) graft the two histories together, rebase, force-push, and have everybody `pull --rebase`) But the final step is always to do a full rebase and freeze all the replacements and grafts into actual commit history, where they will be transmitted when you clone them, without any tricks.

**Comment by rpjday** [comment]
Yes, I see your points, I was just thinking of using a simpler (but not necessarily useful) example of replace to introduce the idea, *then* jumping to the more complicated example.

Oh, and you didn't confirm whether the example currently there could be simplified by just using the "--graft" option instead.

**Comment by ben** [comment]
Probably, yeah. But rather than replacing the example with the shortcut, maybe it would be better to show how it's done using `replace` and _then_ say that it's a common-enough operation to have its own shorthand.
